src/lib/foreign-field.ts

```diff
@@ -0,0 +1,496 @@
+  const p = modulus;
+  const pMl = ForeignFieldConst.fromBigint(p);
+
+  if (p <= 0) {
+    throw Error(`ForeignField: expected modulus to be positive, got ${p}`);
+  }
+  if (p > foreignFieldMax) {
+    throw Error(
+      `ForeignField: modulus exceeds the max supported size of 2^${foreignFieldMaxBits}`
+    );
+  }
+
+  let sizeInBits = p.toString(2).length;
+
+  class ForeignField {
+    static modulus = p;
+    value: ForeignFieldVar;
+
+    /**
+     * Create a new {@link ForeignField} from a bigint, number, string or another ForeignField.
+     *
+     * ```ts
+     * let x = new ForeignField(5);
+     * ```
+     */
+    constructor(x: ForeignField | ForeignFieldVar | bigint | number | string) {
+      if (x instanceof ForeignField) {
+        this.value = x.value;
+        return;
+      }
+      // ForeignFieldVar
+      if (Array.isArray(x)) {
+        this.value = x;
+        return;
+      }
+      // constant
+      this.value = ForeignFieldVar.fromBigint(mod(BigInt(x), p));
+    }
+
+    /**
+     * Coerce the input to a {@link ForeignField}.
+     */
+    static from(x: ForeignField | ForeignFieldVar | bigint | number | string) {
+      if (x instanceof ForeignField) return x;
+      return new ForeignField(x);
+    }
+
+    /**
+     * Checks whether this field element is a constant.
+     *
+     * See {@link FieldVar} tp understand constants vs variables.
+     */
+    isConstant() {
+      let [, ...limbs] = this.value;
+      return limbs.every(FieldVar.isConstant);
+    }
+
+    /**
+     * Convert this field element to a constant.
+     *
+     * See {@link FieldVar} to understand constants vs variables.
+     *
+     * **Warning**: This function is only useful in `Provable.witness()` or `Provable.asProver()` blocks,
+     * that is, in situations where the prover computes a value outside provable code.
+     */
+    toConstant(): ForeignField {
+      let [, ...limbs] = this.value;
+      let constantLimbs = mapTuple(limbs, (l) =>
+        FieldVar.constant(FieldVar.toConstant(l))
+      );
+      return new ForeignField([0, ...constantLimbs]);
+    }
+
+    /**
+     * Convert this field element to a bigint.
+     */
+    toBigInt() {
+      return ForeignFieldVar.toBigint(this.value);
+    }
+
+    /**
+     * Assert that this field element lies in the range [0, p),
+     * where p is the foreign field modulus.
+     */
+    assertValidElement() {
+      if (this.isConstant()) return;
+      Snarky.foreignField.assertValidElement(this.value, pMl);
+    }
+
+    // arithmetic with full constraints, for safe use
+
+    /**
+     * Finite field addition
+     *
+     * ```ts
+     * x.add(2); // x + 2 mod p
+     * ```
+     */
+    add(y: ForeignField | bigint | number) {
+      if (this.isConstant() && isConstant(y)) {
+        let z = mod(this.toBigInt() + toFp(y), p);
+        return new ForeignField(z);
+      }
+      let z = Snarky.foreignField.add(this.value, toVar(y), pMl);
```

Comment:
Thanks for calling that out! (Actually I think `Foreign_field.add` adds only one zero gate)

Yeah, the idea was to land this safe & simple API as a first step, and figure out a more flexible, advanced API in a future PR, after we had the chance to experiment with implementing an optimized gadget ourselves. I think the advanced API will probably live under an `Unsafe` or `Advanced` namespace and expose all possible options.

Providing the `full` flag in the current API didn't feel right to me: the `~full:false` option is not usable by itself without also knowing how to chain additions and without access to the zero gate / the correct bounds check. Apart from that, it makes the full method more error-prone to use. I think `~full:false` is a much lower-level building block that IMO should be a separate method, instead of being mixed with the full `add()` method.

Reply:
Okay! In any case, if you want to provide an easy method to chain, there's the `sum_chain` function which does precisely that. The user basically has to include all of the inputs to be added/subtracted, and a list of operands (+/-), and then it chains it for you, with some final checks at the end.

Reply:
yeah I thought about your comments more and decided I want to include a simple to use method which exposes `sum_chain`! will add that today